float32 a
float32 b
float32 future_x
float32 future_y